### PR TITLE
CacheDictionary request only unique keys from source

### LINKED ISSUE #51762: cache dictionary request the same key lot of times
then one single block of data processed by dictGet have a lot of duplicate keys ClickHouse will construct suboptimal query like 

```sql
SELECT ... FROM dict_source WHERE `id` IN (1000, 1000, ..., 1000, 1000, 1000);
```

instead of simple 

```sql
SELECT ... FROM dict_source WHERE `id` IN (1000);
```

[Repro](https://fiddle.clickhouse.com/bcfe9605-560b-4b85-8c2e-7029fa6f67f5):

```sql
DROP TABLE IF EXISTS dict_source;
DROP DICTIONARY IF EXISTS test_cache_dict;

CREATE TABLE dict_source engine = Log AS SELECT number as id, toString(number) as value FROM numbers(10000);


CREATE DICTIONARY IF NOT EXISTS test_cache_dict (
    id UInt64,
    value String
)
PRIMARY KEY id
SOURCE(
    CLICKHOUSE(
        host '127.0.0.2'
        DB currentDatabase()
        TABLE 'dict_source'
    )
)
LAYOUT(
    CACHE(SIZE_IN_CELLS 10000)
)
LIFETIME(MIN 1 MAX 100);


SELECT dictGet(test_cache_dict, 'value', materialize(toUInt64(1000))) FROM numbers_mt(1000) SETTINGS max_block_size = 50, max_threads = 4 FORMAT Null;

system flush logs;
SELECT event_time, query FROM system.query_log WHERE event_time > now() - 300 and has(tables, currentDatabase() || '.dict_source') and type = 'QueryFinish' and type = 'QueryFinish' and query_kind='Select' ORDER BY event_time DESC LIMIT 10;


SELECT dict_key, count(query_id) number_of_queries_to_source, sum(count_per_query) as sum_key_requests, max(count_per_query) as max_key_requests_per_query FROM (
SELECT  arrayJoin(splitByRegexp(',\s+', extract(query, 'IN \((.*)\);'))) as dict_key, query_id, count() count_per_query FROM system.query_log WHERE event_time > now() - 300 and has(tables, currentDatabase() || '.dict_source') and type = 'QueryFinish' and query_kind='Select' GROUP BY dict_key, query_id) GROUP BY dict_key FORMAT PrettyCompactMonoBlock;
```

### PR DESCRIPTION
**Pull request by kitaisreal**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
CacheDictionary request only unique keys from source. Closes #51762.

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
This is an automated comment for commit 450acb8556f87eec06dda79a83072d357539daea with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/51853/450acb8556f87eec06dda79a83072d357539daea/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

**Review by nickitat — APPROVED** [review]

File: src/Dictionaries/CacheDictionary.cpp [hunk 1]
```diff
@@ -552,16 +544,28 @@ void CacheDictionary<dictionary_key_type>::update(CacheDictionaryUpdateUnitPtr<d
...
+            not_found_key_to_request_index.insert(makePairNoInit(requested_key, i));
         }
     }
 
+    std::vector<UInt64> requested_keys_vector;
+    std::vector<size_t> requested_complex_key_rows;
+
+    if constexpr (dictionary_key_type == DictionaryKeyType::Simple)
+    {
+        requested_keys_vector.reserve(requested_keys.size());
```
Comment (nickitat) [inline src/Dictionaries/CacheDictionary.cpp:558]: not sure if it makes sense, but maybe 
``` cpp
reserve(min(requested_keys.size(), not_found_key_to_request_index.size()))
```
?

Comment (kitaisreal) [reply src/Dictionaries/CacheDictionary.cpp]: Added simpler implementation.

Comment (kitaisreal) [reply src/Dictionaries/CacheDictionary.cpp]: Updated implementation.

**Review by alexey-milovidov — APPROVED** [review]

**Pull request merged**